File: src/75.cpp
#include <algorithm>
#include <iostream>
#include <vector>

using namespace std;

class Solution {
   private:
    void swap(int* n1, int* n2) {
        int tmp = (*n1);
        (*n1) = (*n2);
        (*n2) = tmp;
    }

   public:
    void sortColors(vector<int>& nums) {
        int numsSize = nums.size();
        int ptr = 0;
        for (int i = 0; i < numsSize; ++i) {
            if (nums[i] == 0) {
                swap(&nums[i], &nums[ptr]);
                ptr++;
            }
        }
        for (int i = ptr; i < numsSize; ++i) {
            if (nums[i] == 1) {
                swap(&nums[i], &nums[ptr]);
                ptr++;
            }
        }
    }
};

int main(int argc, char** argv) {
    struct subject {
        vector<int> nums;
    };
    vector<subject> testData{{{2, 0, 2, 1, 1, 0}}, {{2, 0, 1}}};
    int numberOfTestCase = testData.size();
    /* Example
     *  Input: nums = [2,0,2,1,1,0]
     *  Output: [0,0,1,1,2,2]
     *
     *  Input: nums = [2,0,1]
     *  Output: [0,1,2]
     */

    Solution cSolution;
    for (int i = 0; i < numberOfTestCase; ++i) {
        cout << "Input: nums = [";
        for (long unsigned int j = 0; j < testData[i].nums.size(); ++j) {
            cout << (j == 0 ? "" : ",") << testData[i].nums[j];
        }
        cout << "]\n";

        cSolution.sortColors(testData[i].nums);
        cout << "Output: [";
        for (long unsigned int j = 0; j < testData[i].nums.size(); ++j) {
            cout << (j == 0 ? "" : ",") << testData[i].nums[j];
        }
        cout << "]\n";

        cout << "\n";
    }

    return EXIT_SUCCESS;
}

File: src/75.py
import logging
import sys
from bisect import bisect_left, bisect_right
from collections import Counter, defaultdict, deque
from functools import cache, lru_cache
from heapq import heapify, heappop, heappush
from itertools import accumulate
from typing import List, Optional


def logging_setting():
    LOG_LEVEL = logging.INFO    # DEBUG, INFO, WARNING, ERROR, CRITICAL
    if __debug__:
        LOG_LEVEL = logging.DEBUG

    logging.basicConfig(
        stream=sys.stderr,
        level=LOG_LEVEL,
        format="%(levelname)-6s %(asctime)s %(module)s:%(funcName)s:%(lineno)-4s %(message)s",
        datefmt='%Y/%m/%d %H:%M:%S')


class Solution:
    def sortColors(self, nums: List[int]) -> None:
        """
        Do not return anything, modify nums in-place instead.
        """
        numsSize = len(nums)
        ptr = 0
        for i in range(numsSize):
            if nums[i] == 0:
                nums[i], nums[ptr] = nums[ptr], nums[i]
                ptr += 1
        for i in range(ptr, numsSize):
            if nums[i] == 1:
                nums[i], nums[ptr] = nums[ptr], nums[i]
                ptr += 1


if __name__ == "__main__":
    logging_setting()

    try:
        logging.info("sys.version: %s", sys.version)
        print()

        pSolution = Solution()
        for nums in [[2, 0, 2, 1, 1, 0], [2, 0, 1]]:
            # /* Example
            #  *  Input: nums = [2,0,2,1,1,0]
            #  *  Output: [0,0,1,1,2,2]
            #  *
            #  *  Input: nums = [2,0,1]
            #  *  Output: [0,1,2]
            #  */
            logging.info("Input: nums = %s", nums)

            pSolution.sortColors(nums)
            logging.info("Output: %s", nums)

            print()
    except KeyboardInterrupt as exception:
        logging.error("%s: %s", exception.__class__.__name__, exception, exc_info=True)
        pass
    except Exception as exception:
        logging.error("%s: %s", exception.__class__.__name__, exception, exc_info=True)
        pass

    sys.exit(0)
